src/Raven.Client/Documents/BulkInsert/BulkInsertOperation.cs

```diff
@@ -305,76 +295,184 @@ public async Task StoreAsync(object entity, string id, IMetadataDictionary metad
                         metadata[Constants.Documents.Metadata.RavenClrType] = clrType;
                 }
 
-                if (_first == false)
-                {
-                    _currentWriter.Write(',');
-                }
-
-                _first = false;
-                try
+                await WriteToStream(writeComma: true, () =>
                 {
                     _currentWriter.Write("{\"Id\":\"");
-                    WriteId(_currentWriter, id);
+                    WriteString(_currentWriter, id);
                     _currentWriter.Write("\",\"Type\":\"PUT\",\"Document\":");
 
                     if (_customEntitySerializer == null || _customEntitySerializer(entity, metadata, _currentWriter) == false)
                     {
                         using (var json = EntityToBlittable.ConvertEntityToBlittable(entity, _conventions, _context,
-                            _defaultSerializer, new DocumentInfo
-                            {
-                                MetadataInstance = metadata
-                            }))
-
+                            _defaultSerializer, new DocumentInfo {MetadataInstance = metadata}))
                         {
                             _currentWriter.Flush();
                             json.WriteJsonTo(_currentWriter.BaseStream);
                         }
                     }
 
                     _currentWriter.Write("}");
-                    _currentWriter.Flush();
+                }, id, CommandType.PUT).ConfigureAwait(false);
+            }
+        }
 
-                    if (_currentWriter.BaseStream.Position > _maxSizeInBuffer ||
-                        _asyncWrite.IsCompleted)
-                    {
-                        await _asyncWrite.ConfigureAwait(false);
-
-                        var tmp = _currentWriter;
-                        _currentWriter = _backgroundWriter;
-                        _backgroundWriter = tmp;
-                        _currentWriter.BaseStream.SetLength(0);
-                        ((MemoryStream)tmp.BaseStream).TryGetBuffer(out var buffer);
-                        _asyncWrite = _requestBodyStream.WriteAsync(buffer.Array, buffer.Offset, buffer.Count, _token);
-                    }
+        public void AppendTimeSeries(string id, string timeseries, DateTime timestamp, string tag, IEnumerable<double> values)
+        {
+            AsyncHelpers.RunSync(() => AppendTimeSeriesAsync(id, timeseries, timestamp, tag, values));
+        }
+
+        public async Task AppendTimeSeriesAsync(string id, string timeseries, DateTime timestamp, string tag, IEnumerable<double> values)
```

Comment:
Wondering if it would not be better to have 

```csharp
bulkInsert.TimeSeries.Append
bulkInsert.Counters.Increment
```

We are doing similar separation in session.

Reply:
I like that. Especially true since we need to have better API so we can have:

`bulk.TimeSeriesFor(docId, name)` And then you call `Append` multiple times here.
That matches the underlying format.

Reply:
We can either do it now or as a part of TimeSeries API revamp. @karmeli87 @aviv86 ?

Reply:
I like Oren's suggestion and would like to have it in the `session`, `bulkInsert` and `patching` and adjust how we send stuff on the wire accordingly.
 
We will handle this in a separate issue as part of the API unification.